File: CSharp/MNG.cs
using System;
using System.Collections.Generic;

namespace Mng
{
    public sealed class MngFrame
    {
        public MngFrame(byte[] png, uint delayMilliseconds)
        {
            Png = png;
            DelayMilliseconds = delayMilliseconds;
        }

        public byte[] Png { get; }
        public uint DelayMilliseconds { get; }
    }

    public sealed class MngAnimation
    {
        public uint Width { get; internal set; }
        public uint Height { get; internal set; }
        public uint TicksPerSecond { get; internal set; } = 100;
        public uint NominalLayerCount { get; internal set; }
        public uint NominalFrameCount { get; internal set; }
        public uint NominalPlayTime { get; internal set; }
        public uint SimplicityProfile { get; internal set; }
        public IList<MngFrame> Frames { get; } = new List<MngFrame>();

        public int FrameIndexAt(TimeSpan elapsed)
        {
            if (Frames.Count == 0) return -1;
            if (Frames.Count == 1) return 0;
            long duration = 0;
            foreach (MngFrame frame in Frames) duration += frame.DelayMilliseconds;
            if (duration <= 0) return 0;
            long milliseconds = (long)elapsed.TotalMilliseconds % duration;
            if (milliseconds < 0) milliseconds += duration;
            for (int index = 0; index < Frames.Count; index++)
            {
                if (milliseconds < Frames[index].DelayMilliseconds) return index;
                milliseconds -= Frames[index].DelayMilliseconds;
            }
            return Frames.Count - 1;
        }
    }

    public static class MngParser
    {
        private static readonly byte[] MngSignature = { 0x8a, 0x4d, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a };
        private static readonly byte[] PngSignature = { 0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a };

        public static bool IsMng(byte[] data)
        {
            if (data == null || data.Length < MngSignature.Length) return false;
            for (int index = 0; index < MngSignature.Length; index++) if (data[index] != MngSignature[index]) return false;
            return true;
        }

        public static MngAnimation Parse(byte[] data)
        {
            if (!IsMng(data)) throw new FormatException("not an MNG image");
            var animation = new MngAnimation();
            var frame = new List<byte>();
            uint pendingDelay = 0;
            bool hasPendingDelay = false;
            bool inPng = false;
            int offset = MngSignature.Length;
            while (offset < data.Length)
            {
                if (data.Length - offset < 12) throw new FormatException("truncated MNG chunk");
                ulong dataLength = ReadU32(data, offset);
                ulong chunkLength = dataLength + 12;
                if (chunkLength < 12 || chunkLength > (ulong)(data.Length - offset)) throw new FormatException("invalid MNG chunk length");
                int length = (int)dataLength;
                string type = System.Text.Encoding.ASCII.GetString(data, offset + 4, 4);
                if (type == "MHDR")
                {
                    if (length < 28) throw new FormatException("invalid MNG header");
                    animation.Width = ReadU32(data, offset + 8);
                    animation.Height = ReadU32(data, offset + 12);
                    uint ticks = ReadU32(data, offset + 16);
                    if (ticks > 0) animation.TicksPerSecond = ticks;
                    animation.NominalLayerCount = ReadU32(data, offset + 20);
                    animation.NominalFrameCount = ReadU32(data, offset + 24);
                    animation.NominalPlayTime = ReadU32(data, offset + 28);
                    animation.SimplicityProfile = ReadU32(data, offset + 32);
                }
                else if (type == "FRAM") hasPendingDelay = ParseFrameDelay(data, offset + 8, length, out pendingDelay);
                if (!inPng && type == "IHDR")
                {
                    frame.AddRange(PngSignature);
                    inPng = true;
                }
                if (inPng)
                {
                    for (int index = 0; index < (int)chunkLength; index++) frame.Add(data[offset + index]);
                    if (type == "IEND")
                    {
                        uint delay = 100;
                        if (hasPendingDelay)
                        {
                            ulong milliseconds = ((ulong)pendingDelay * 1000 + animation.TicksPerSecond / 2) / animation.TicksPerSecond;
                            delay = milliseconds > uint.MaxValue ? uint.MaxValue : (uint)milliseconds;
                        }
                        animation.Frames.Add(new MngFrame(frame.ToArray(), delay));
                        frame.Clear();
                        pendingDelay = 0;
                        hasPendingDelay = false;
                        inPng = false;
                    }
                }
                offset += (int)chunkLength;
            }
            if (inPng || animation.Frames.Count == 0) throw new FormatException("MNG contains no complete frames");
            return animation;
        }

        public static byte[] FirstPngFrame(byte[] data)
        {
            return (byte[])Parse(data).Frames[0].Png.Clone();
        }

        private static uint ReadU32(byte[] data, int offset)
        {
            return ((uint)data[offset] << 24) | ((uint)data[offset + 1] << 16) | ((uint)data[offset + 2] << 8) | data[offset + 3];
        }

        private static bool ParseFrameDelay(byte[] data, int offset, int length, out uint delay)
        {
            delay = 0;
            if (length == 0) return false;
            int index = 1;
            while (index < length && data[offset + index] != 0) index++;
            if (index == length) return false;
            index++;
            if (index >= length) return false;
            bool hasDelay = data[offset + index++] != 0;
            if (length - index < 3) return false;
            index += 3;
            if (!hasDelay || length - index < 4) return false;
            delay = ReadU32(data, offset + index);
            return true;
        }
    }
}

File: Python/mng.py
from dataclasses import dataclass, field
from datetime import timedelta
from typing import List, Optional, Union

MNG_SIGNATURE = bytes((0x8A, 0x4D, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A))
PNG_SIGNATURE = bytes((0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A))

class MNGError(ValueError):
    pass

@dataclass(frozen=True)
class MNGFrame:
    png: bytes
    delay_milliseconds: int = 100

@dataclass
class MNGAnimation:
    width: int = 0
    height: int = 0
    ticks_per_second: int = 100
    nominal_layer_count: int = 0
    nominal_frame_count: int = 0
    nominal_play_time: int = 0
    simplicity_profile: int = 0
    frames: List[MNGFrame] = field(default_factory=list)

    def frame_index_at(self, elapsed: Union[timedelta, int]) -> int:
        if not self.frames:
            return -1
        if len(self.frames) == 1:
            return 0
        milliseconds = int(elapsed.total_seconds() * 1000) if isinstance(elapsed, timedelta) else int(elapsed)
        duration = sum(frame.delay_milliseconds for frame in self.frames)
        if duration <= 0:
            return 0
        elapsed = milliseconds % duration
        for index, frame in enumerate(self.frames):
            if elapsed < frame.delay_milliseconds:
                return index
            elapsed -= frame.delay_milliseconds
        return len(self.frames) - 1

def is_mng(data: bytes) -> bool:
    return data[:len(MNG_SIGNATURE)] == MNG_SIGNATURE

def _read_u32(data: bytes, offset: int) -> int:
    return int.from_bytes(data[offset:offset + 4], "big")

def _parse_frame_delay(data: bytes) -> Optional[int]:
    if not data:
        return None
    offset = 1
    while offset < len(data) and data[offset] != 0:
        offset += 1
    if offset == len(data):
        return None
    offset += 1
    if offset >= len(data):
        return None
    has_delay = data[offset] != 0
    offset += 1
    if len(data) - offset < 3:
        return None
    offset += 3
    if not has_delay or len(data) - offset < 4:
        return None
    return _read_u32(data, offset)

def parse(data: bytes) -> MNGAnimation:
    if not is_mng(data):
        raise MNGError("not an MNG image")
    animation = MNGAnimation()
    frame = bytearray()
    pending_delay: Optional[int] = None
    in_png = False
    offset = len(MNG_SIGNATURE)
    while offset < len(data):
        if len(data) - offset < 12:
            raise MNGError("truncated MNG chunk")
        data_length = _read_u32(data, offset)
        chunk_length = data_length + 12
        if chunk_length < 12 or chunk_length > len(data) - offset:
            raise MNGError("invalid MNG chunk length")
        chunk_type = data[offset + 4:offset + 8]
        chunk_data = data[offset + 8:offset + 8 + data_length]
        if chunk_type == b"MHDR":
            if data_length < 28:
                raise MNGError("invalid MNG header")
            animation.width = _read_u32(chunk_data, 0)
            animation.height = _read_u32(chunk_data, 4)
            ticks = _read_u32(chunk_data, 8)
            if ticks > 0:
                animation.ticks_per_second = ticks
            animation.nominal_layer_count = _read_u32(chunk_data, 12)
            animation.nominal_frame_count = _read_u32(chunk_data, 16)
            animation.nominal_play_time = _read_u32(chunk_data, 20)
            animation.simplicity_profile = _read_u32(chunk_data, 24)
        elif chunk_type == b"FRAM":
            pending_delay = _parse_frame_delay(chunk_data)
        if not in_png and chunk_type == b"IHDR":
            frame = bytearray(PNG_SIGNATURE)
            in_png = True
        if in_png:
            frame.extend(data[offset:offset + chunk_length])
            if chunk_type == b"IEND":
                delay = 100 if pending_delay is None else round(pending_delay * 1000 / animation.ticks_per_second)
                animation.frames.append(MNGFrame(bytes(frame), max(0, delay)))
                frame.clear()
                pending_delay = None
                in_png = False
        offset += chunk_length
    if in_png or not animation.frames:
        raise MNGError("MNG contains no complete frames")
    return animation

def first_png_frame(data: bytes) -> bytes:
    return parse(data).frames[0].png
